source/configuration/modules/omfile.rst

```diff
@@ -1,434 +1,441 @@
-   rsyslogd on during startup processing. Interim changes to the user
-   mapping are not detected.
-
-.. function::  dirGroupNum [gid]
-
-   *Default: system default*
-
-   Set the group for directories newly created. Please note that this
-   setting does not affect the group of directories already existing.
-   The parameter is a numerical ID, which is used regardless of whether
-   the group actually exists. This can be useful if the group mapping is
-   not available to rsyslog during startup.
-
-.. function::  fileOwner [userName]
-
-   *Default: system default*
-
-   Set the file owner for files newly created. Please note that this
-   setting does not affect the owner of files already existing. The
-   parameter is a user name, for which the userid is obtained by
-   rsyslogd during startup processing. Interim changes to the user
-   mapping are *not* detected.
-
-.. function::  fileOwnerNum [uid]
-
-   *Default: system default*
-
-   *Available since: 7.5.8, 8.1.4*
-
-   Set the file owner for files newly created. Please note that this
-   setting does not affect the owner of files already existing. The
-   parameter is a numerical ID, which which is used regardless of
-   whether the user actually exists. This can be useful if the user
-   mapping is not available to rsyslog during startup.
+People often report problems that dynafiles are not properly created.
+Apart from generating invalid paths/filenames, sometimes these problems are due to SELinux rules, which do not permit
+the creation of those files. The same happens for generic permission issues (this is often a problem
+under Ubuntu where permissions are dropped by default)
+
+=========  ==========================
+template:  "template-name"
+Default:   none
+Legacy:    $ActionFileDefaultTemplate
+=========  ==========================
+
+Sets the template to be used for this action.
+
+================  ==============================
+closeTimeout:     "number"
+Default:          "0 if using *file*, 10 if using *dynaFile*"
+Available since:  8.3.3
+================  ==============================
+
+Specifies after how many minutes of inactivity a file is
+automatically closed.
+
+Note that this functionality is implemented based on the  :doc:`janitor process <../../concepts/janitor>`.
+
+==================  ==================
+dynaFileCacheSize:  "number"
+Default:            "10"
+Legacy:             $DynaFileCacheSize
+==================  ==================
+
+**Applies only if dynamic filenames are used.**
+
+Specifies the number of files that will be kept open.
+Note that this is a per-action parameter, so if you have
+multiple dynafile actions, each of them have their individual caches
+(which means the numbers sum up).
+
+Ideally, the cache size exactly
+matches the need. You can use :doc:`impstats <impstats>` to tune
+this value. Note that a too-low cache size can be a very considerable
+performance bottleneck.
+
```

Comment:
it's far better to have the cache be too large than too small. Even being too small by 1 file can cripple your performance.

If the cache is too small, every write turns into:

1. find the oldest file
2. flush all data in buffers for this oldest file
3. close the oldest file
4. open the new file
5. write the data into the buffers for the new file